drivers/adc/adc_mcux_lpadc.c

```diff
@@ -461,13 +479,164 @@ static void mcux_lpadc_start_channel(const struct device *dev)
 	LPADC_DoSoftwareTrigger(config->base, 1);
 }
 
+#ifdef CONFIG_LPADC_ENABLE_DMA
+static void mcux_lpadc_dma_callback(const struct device *dma_dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct mcux_lpadc_data *data = (struct mcux_lpadc_data *)user_data;
+	const struct device *dev = data->dev;
+	const struct mcux_lpadc_config *config = dev->config;
+
+	if (!data->use_dma) {
+		return;
+	}
+
+	(void)dma_stop(config->dma_dev, config->dma_channel);
+
+	if (status < 0) {
+		adc_context_complete(&data->ctx, status);
+		return;
+	}
+
+	/* Disable LPADC DMA request */
+#if (defined(FSL_FEATURE_LPADC_FIFO_COUNT) && (FSL_FEATURE_LPADC_FIFO_COUNT == 2U))
+	LPADC_EnableFIFO0WatermarkDMA(config->base, false);
+#else
+	LPADC_EnableFIFOWatermarkDMA(config->base, false);
+#endif
+
+	/* Process the data in staging buffer and pass to data->buffer in enabled channel order. */
+	uint32_t written = 0U;
+
+	for (uint8_t ch = 0U;
+		ch < CONFIG_LPADC_CHANNEL_COUNT && written < data->channels_count; ch++) {
+		/* Skip channels not requested in this sequence. */
+		if (!(data->ctx.sequence.channels & BIT(ch))) {
+			continue;
+		}
+
+		lpadc_sample_channel_mode_t conv_mode = data->cmd_config[ch].sampleChannelMode;
+
+		uint32_t fifo_data = data->dma_results[written];
+		uint16_t conv_result = (uint16_t)(fifo_data & 0xFFFF);
+		uint16_t tmp;
+		bool is_diff = false;
+
+#if defined(FSL_FEATURE_LPADC_HAS_B_SIDE_CHANNELS) && (FSL_FEATURE_LPADC_HAS_B_SIDE_CHANNELS)
+#if defined(FSL_FEATURE_LPADC_HAS_CMDL_DIFF) && (FSL_FEATURE_LPADC_HAS_CMDL_DIFF)
+		is_diff = (conv_mode == kLPADC_SampleChannelDiffBothSideAB ||
+			conv_mode == kLPADC_SampleChannelDiffBothSideBA);
+#else
+		is_diff = (conv_mode == kLPADC_SampleChannelDiffBothSide);
+#endif
+#endif
+
+		/* Format handling:
+		 * - 16-bit differential: bit 15 sign; bit [0-14] data (two's complement).
+		 * - 16-bit single-ended: bit [0-15] data (unsigned).
+		 * - 13-bit differential: bit 15 sign; bit [3-14] data; bit [0-2] zero.
+		 * - 12-bit single-ended: bit [3-14] data; bit [0-2] zero.
+		 */
+		if (data->ctx.sequence.resolution == 16) {
+			/* Keep raw 16-bit value; for differential, bit15 is sign. */
+			tmp = conv_result;
+		} else {
+			/* Right-align 12/13-bit payload from bit [3-14]. */
+			uint16_t value = (uint16_t)((conv_result >> 3) & 0xFFF);
+
+			if (is_diff && (conv_result & 0x8000)) {
+				/* 13-bit differential: apply two's complement sign
+				 * using 12-bit magnitude.
+				 */
+				value = (uint16_t)(value - 0x1000);
+			}
+			tmp = value;
+		}
+
+		*data->buffer++ = tmp;
+		written++;
+	}
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+}
+
+/* Configure and start DMA before triggering conversions */
+static void mcux_lpadc_dma_configure(struct mcux_lpadc_data *data)
+{
+	const struct device *dev = data->dev;
+	const struct mcux_lpadc_config *config = dev->config;
+
+	if (data->use_dma) {
```

Comment:
reverse the condition to if (!data->use_dma) { return; } to remove a level of nesting /cognitive complexity